frontend/resources/include/GenericImage.h

```diff
@@ -0,0 +1,160 @@
+#include <string>
+#include <optional>
+#include <functional>
+
+namespace megamol {
+namespace frontend_resources {
+
+enum class GenericImageType {
+    GLTexureHandle, // data array holds a GL texture handle
+    ByteArray,      // data array holds image data as raw bytes
+    BorrowedGLTexureHandle, // holds GL texture handle, but non-owning
+    BorrowedByteArray       // holds image data as raw bytes, but non-owning
+};
+static const GenericImageType GLTexureHandle = GenericImageType::GLTexureHandle;
+static const GenericImageType ByteArray = GenericImageType::ByteArray;
+static const GenericImageType BorrowedGLTexureHandle = GenericImageType::BorrowedGLTexureHandle;
+static const GenericImageType BorrowedByteArray = GenericImageType::BorrowedByteArray;
+
+// the idea is that each AbstractView (via the concrece View implementation)
+// fills a GenericImage for the frontend to use.
+// depending on the GenericImageType and DataChannels the frontend knows
+// what it can do with each GenericImage
+//   - either use the contained texture directly for GL rendering
+//   - or forward the contained byte data to interested sources: show in window, write to screenshot file, send via network...
+// we expect the content type and channel count to stay the same across the lifetime of the corresponding source View
+// so after a view finishes rendering, it writes the final FBO state into a GenericImage
+struct GenericImage {
+
+    enum class DataChannels {
+        // for texture and byte array, tells us how many channels there are
+        RGB8,
+        RGBA8,
+    };
+    struct ImageSize {
+        size_t width = 0;
+        size_t height = 0;
+
+        bool operator!=(ImageSize const& other) { return width != other.width || height != other.height; }
+    };
+
+    GenericImage(ImageSize size, DataChannels channels, GenericImageType type, const void* data);
+   
+    GenericImage() = default;
+    ~GenericImage(); // GL texture deletion
+    // rule of five
+    GenericImage(GenericImage const& other);
+    GenericImage& operator=(GenericImage const& other);
+    GenericImage(GenericImage&& other) noexcept;
+    GenericImage& operator=(GenericImage&& other) noexcept;
+
+    GenericImageType image_type{};
+    ImageSize image_size{};
+    DataChannels channels{};
+    bool new_data = false;
+    bool new_size = false;
+
+    using byte = unsigned char;
+    // if byte image, image_data holds (width * height * channels) entries.
+    // row-major image starting at bottom-left pixel
+    // as in https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glReadPixels.xhtml
+    mutable std::vector<byte> image_data; // mutable for texture download
+
+    // note that gl.h defines "typedef unsigned int GLuint;"
+    mutable unsigned int gl_texture_handle = 0; // mutable for texture upload
+    void download_texture() const; // downloads texture into local image_data
+    void upload_texture() const; // downloads texture into local image_data
```

Comment:
Guess the comment is not accurate? This one uploads texture data from local image data, right?